crash fixes and updates to client.py

clanmcgoon opened the pull request:
1.The client would crash if no server connection was established. This has been fixed.
2.The Chat interface was not working as intended. Pressing the [t] key was meant to open up the chat but was not properly setting the focus to the chat entry. This has been fixed and the code for placement of the chat has been changed to pin it to the lower left corner regardless of screen resolution.
3.The camera was, in my opinion, not practical. The code for moving the camera has been replaced with code that allows the user to toggle between 3 camera views. The default is a third person shoulder level PoV, the second is a first person PoV, and the third is a free camera PoV ( hold RMB to zoom in and out, hold MMB to change angle of view, press [c] key to toggle between views ).
3a. Auto Run is now available by pressing the period key [.] on either the standard keyboard or the numpad.

Commit:
crash fixes and updates to client.py

1.The client would crash if no server connection was established. This has been fixed.
2.The Chat interface was not working as intended. Pressing the [t] key was meant to open up the chat but was not properly setting the focus to the chat entry. This has been fixed and the code for placement of the chat has been changed to pin it to the lower left corner regardless of screen resolution.
3.The camera was, in my open, not practical. The code for moving the camera has been replaced with code that allows the user to toggle between 3 camera views. The default is a third person shoulder level PoV, the second is a first person PoV, and the third is a free camera PoV ( hold RMB to zoom in and out, hold MMB to change angle of view, press [c] key to toggle between views ).
3a. Auto Run is now available by pressing the period key [.] on either the standard keyboard or the numpad.

clanmcgoon commented:
Sorry about the indents as I could not easily change from my 2 indent style to the 8 indent style.

brthornbury commented on client/client.py:
```diff
@@ -9,295 +9,375 @@
+  def __init__(self):
+    self.terrain = GeoMipTerrain("mySimpleTerrain")
+    self.terrain.setHeightfield(Filename("heightmap.png"))
+    self.terrain.setColorMap(Filename("terrain.bmp"))  #pjb comment this line out if you want to set texture directly
```
What is pjb here?

clanmcgoon replied:
pjb those are my initials so marking it as a personal note

brthornbury commented on client/client.py:
```diff
@@ -9,295 +9,375 @@
+  def move(self, keyClass, terrainClass):
+    #self.meTerrainHeight = terrainClass.terrain.getElevation(self.model.getX(),self.model.getY()) * self.terrainScale
+    #self.camTerrainHeight = terrainClass.terrain.getElevation(camera.getX(),camera.getY()) * self.terrainScale
+    self.camDummy = self.model.attachNewNode("camDummy")
```
Is it possible to reuse the node rather than reattaching every time?

clanmcgoon replied:
absolutely. It should be part of the Init process. That is a good thing you caught this as I was noticing a memory leak and that was probably causing it. I made the cam dummy for the camera to point at instead of the head because the head bobs up and down when animated causing the camera to shake up and down in a very unpleasant manner. I can make the changes and then resubmit a new PR or what ever you think is best.

Mozoby commented:
Very nice! Thanks for taking the time to make the PR :smile: 

Had a couple of questions, but overall looks good.

Commit:
Patch fix

Changed the self.camDummy node creation from the move() method of the me class to the class init() method.

Commit:
the file name was wrong causing a crash

heightmap.png changed to Heightmap.png

clanmcgoon commented:
Sorry for the additional commits. This Pull Request should be good to go.

Mozoby commented:
Nice!
